pallets/identity/src/keys.rs

```diff
@@ -168,24 +168,37 @@ impl<T: Config> Module<T> {
 
     /// Checks that a key is not linked to any identity or multisig.
     pub fn can_link_account_key_to_did(key: &T::AccountId) -> bool {
-        !<KeyToIdentityIds<T>>::contains_key(key) && !T::MultiSig::is_signer(key)
+        !KeyRecords::<T>::contains_key(key) && !T::MultiSig::is_signer(key)
     }
 
     /// Links a primary or secondary `AccountId` key `key` to an identity `did`.
     ///
     /// This function applies the change if `can_link_account_key_to_did` returns `true`.
     /// Otherwise, it does nothing.
-    pub fn link_account_key_to_did(key: &T::AccountId, did: IdentityId) {
-        if !<KeyToIdentityIds<T>>::contains_key(key) {
+    pub fn link_account_key_to_did(key: &T::AccountId, record: KeyRecord) {
+        if !KeyRecords::<T>::contains_key(key) {
             // `key` is not yet linked to any identity, so no constraints.
-            <KeyToIdentityIds<T>>::insert(key, did);
+            KeyRecords::<T>::insert(key, &record);
+            DidKeys::<T>::insert(record.did, key, true);
+            if record.is_primary_key() {
+                DidPrimaryKey::<T>::insert(record.did, DidRecord::new(key.clone()));
+            }
         }
     }
 
     /// Unlinks an `AccountId` key `key` from an identity `did`.
     fn unlink_account_key_from_did(key: &T::AccountId, did: IdentityId) {
-        if <KeyToIdentityIds<T>>::contains_key(key) && <KeyToIdentityIds<T>>::get(key) == did {
-            <KeyToIdentityIds<T>>::remove(key)
+        if DidKeys::<T>::take(did, key) {
+            let is_primary_key = KeyRecords::<T>::take(key).map_or(false, |r| r.is_primary_key());
+            if is_primary_key {
+                DidPrimaryKey::<T>::mutate(did, |d| {
+                    if let Some(ref mut d) = d {
+                        if d.primary_key.as_ref() == Some(key) {
+                            d.primary_key = None;
+                        }
+                    }
```

Comment:
```suggestion
                    *d = Some(DidRecord { primary_key: None });
```
It seems to me that there's some data integrity problem if these checks fail, so we can just assume they don't and overwrite with the end result?

Reply:
`DidRecord` will be removed.  Like I said this PR is not ready for reviews.